packages/block-library/src/navigation/view.js

```diff
@@ -73,60 +73,77 @@ store( {
+					return context.core.navigation.type === 'overlay' &&
+						selectors.core.navigation.isMenuOpen( store )
 						? 'dialog'
-						: '',
+						: '';
+				},
 				isMenuOpen: ( { context } ) =>
-					// The menu is opened if either `click` or `hover` is true.
-					Object.values( context.core.navigation.isMenuOpen ).filter(
-						Boolean
-					).length > 0,
+					// The menu is opened if either `click`, `hover` or `focus` is true.
+					Object.values(
+						context.core.navigation[
+							context.core.navigation.type === 'overlay'
+								? 'overlayOpenedBy'
+								: 'submenuOpenedBy'
+						]
+					).filter( Boolean ).length > 0,
+				menuOpenBy: ( { context } ) =>
+					context.core.navigation[
+						context.core.navigation.type === 'overlay'
+							? 'overlayOpenedBy'
+							: 'submenuOpenedBy'
+					],
 			},
 		},
 	},
 	actions: {
 		core: {
 			navigation: {
-				openMenuOnHover( args ) {
-					openMenu( args, 'hover' );
+				openMenuOnHover( store ) {
+					const { navigation } = store.context.core;
+					if (
+						navigation.type === 'submenu' &&
+						// Only open on hover if the overlay is closed.
+						Object.values(
+							navigation.overlayOpenedBy || {}
+						).filter( Boolean ).length === 0
+					)
+						openMenu( store, 'hover' );
 				},
-				closeMenuOnHover( args ) {
-					closeMenu( args, 'hover' );
+				closeMenuOnHover( store ) {
+					closeMenu( store, 'hover' );
 				},
-				openMenuOnClick( args ) {
-					openMenu( args, 'click' );
+				openMenuOnClick( store ) {
+					openMenu( store, 'click' );
 				},
-				closeMenuOnClick( args ) {
-					closeMenu( args, 'click' );
+				closeMenuOnClick( store ) {
+					closeMenu( store, 'click' );
 				},
-				toggleMenuOnClick: ( args ) => {
-					const { context } = args;
-					if ( context.core.navigation.isMenuOpen.click ) {
-						closeMenu( args, 'click' );
+				toggleMenuOnClick: ( store ) => {
+					const { selectors } = store;
+					if ( selectors.core.navigation.menuOpenBy( store ).click ) {
+						closeMenu( store, 'click' );
 					} else {
-						openMenu( args, 'click' );
+						openMenu( store, 'click' );
 					}
 				},
-				handleMenuKeydown: ( args ) => {
-					const { context, event } = args;
-					if ( context.core.navigation.isMenuOpen.click ) {
-						// If Escape close the menu
-						if (
-							event?.key === 'Escape' ||
-							event?.keyCode === 27
-						) {
-							closeMenu( args, 'click' );
+				handleMenuKeydown: ( store ) => {
+					const { context, selectors, event } = store;
+					if ( selectors.core.navigation.menuOpenBy( store ).click ) {
+						// If Escape close the menu.
+						if ( event?.key === 'Escape' ) {
+							closeMenu( store, 'click' );
 							return;
 						}
 
-						// Trap focus if it is an overlay (main menu)
+						// Trap focus if it is an overlay (main menu).
 						if (
-							context.core.navigation.overlay &&
-							( event.key === 'Tab' || event.keyCode === 9 )
```

Comment:
I removed the `keyCode`s because they are deprecated and not needed in modern browsers: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode